spectec/spec/7-module.watsup

```diff
@@ -0,0 +1,134 @@
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (TABLE x)) = { NAME name, VALUE (TABLE ta*[x]) }
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (MEM x)) = { NAME name, VALUE (MEM ma*[x]) }
+
+
+def $allocfunc(moduleinst, func) : funcinst
+def $allocfunc(m, func) = { MODULE m, CODE func }
+
+def $allocglobal(globaltype, val) : globalinst
+def $allocglobal(globaltype, val) = { TYPE globaltype, VALUE val }
+
+def $alloctable(tabletype) : tableinst
+def $alloctable(`[i .. j] rt) = { TYPE (`[i .. j] rt), ELEM (REF.NULL rt)^i }
+
+def $allocmem(memtype) : meminst
+def $allocmem(`[i .. j] I8) = { TYPE (`[i .. j] I8), DATA 0^(i * 64 * $Ki()) }
+
+def $allocelem(reftype, ref*) : eleminst
+def $allocelem(rt, ref*) = { TYPE rt, ELEM ref* }
+
+def $allocdata(byte*) : datainst
+def $allocdata(byte*) = { DATA byte* }
+
+
+def $allocmodule(store, module, externval*, val*, (ref*)*) : (store, moduleinst)
+def $allocmodule(s, module, externval*, val*, (ref*)*) = (s_6, m)
+  -- if module =
+    MODULE
+      import*
+      func*
+      (GLOBAL globaltype expr)*
+      (TABLE tabletype)*
+      (MEMORY memtype)*
+      (ELEM rt expr* elemmode?)*
+      (DATA byte* datamode?)*
+      start?
+      export*
+  -- if fa_ex* = $funcs(externval*)
+  -- if ga_ex* = $globals(externval*)
+  -- if ta_ex* = $tables(externval*)
+  -- if ma_ex = $mems(externval*)
+  -- if fa* = |s.FUNC| ... $(|s.FUNC| + |func*| - 1)
+  -- if ga* = |s.GLOBAL| ... $(|s.GLOBAL| + |global*| - 1)
+  -- if ta* = |s.TABLE| ... $(|s.TABLE| + |table*| - 1)
+  -- if ma* = |s.MEM| ... $(|s.MEM| + |mem*| - 1)
+  -- if ea* = |s.ELEM| ... $(|s.ELEM| + |elem*| - 1)
+  -- if da* = |s.DATA| ... $(|s.DATA| + |data*| - 1)
+  -- if xi* = $instexport(fa_ex* fa*, ga_ex* ga*, ta_ex* ta*, ma_ex* ma*, export)*
+  -- if m = {
+      FUNC fa_ex* fa*,
+      GLOBAL ga_ex* ga*,
+      TABLE ta_ex* ta*,
+      MEM ma_ex* ma*,
+      ELEM ea*,
+      DATA da*,
+      EXPORT xi*
+    }
+  -- if s_1 = s[.FUNC =.. $allocfunc(m, func)*]
+  -- if s_2 = s_1[.GLOBAL =.. $allocglobal(globaltype, val)*]
+  -- if s_3 = s_2[.TABLE =.. $alloctable(tabletype)*]
+  -- if s_4 = s_3[.MEM =.. $allocmem(memtype)*]
+  -- if s_5 = s_4[.ELEM =.. $allocelem(rt, ref*)*]
+  -- if s_6 = s_5[.DATA =.. $allocdata(byte*)*]
+
+
+def $runelem(state, elem, idx) : instr*
+def $runelem((s; f), ELEM reftype expr*, i) = epsilon
+def $runelem((s; f), ELEM reftype expr* (DECLARE), i) = (ELEM.DROP i)
+def $runelem((s; f), ELEM reftype expr* (TABLE x instr*), i) =
+  instr* (CONST I32 0) (CONST I32 n) (TABLE.INIT x i) (ELEM.DROP i)
+  -- if n = |expr*|
+
+def $rundata(state, data, idx) : instr*
+def $rundata((s; f), DATA byte*, i) = epsilon
+def $rundata((s; f), DATA byte* (MEMORY 0 instr*), i) =
+  instr* (CONST I32 0) (CONST I32 n) (MEMORY.INIT i) (DATA.DROP i)
+  -- if n = |byte*|
+
+
+def $instantiation(store, module, externval*) : config
+def $instantiation(s, module, externval*) =
+  s'; f; ...$runelem(elem*[i], i)* ...$rundata(data*[j], j)* (CALL x)?
```

Comment:
We also need range for `runelem` and `rundata`.
Here, `$runelem(elem*[i], i)*` means `$runelem(elem*[0], 0) ... $runelem(elem*[n-1]), n-1)` where `i* = 0 ... $(n - 1)`.
However, return type of `runelem` is `instr*`, so return type of `$runelem(..)*` is `instr**`.
Therefore, it requires new syntax for flattening.

`...` notation comes from JavaScript syntax spread.
Alternatively, we can use notations from Scala or Python
Scala repeated parameters: `$runelem(..):_*`
Python unpacking: `*($runelem(..)*)`

<img width="532" alt="image" src="https://github.com/Wasm-DSL/spectec/assets/50018375/19001bab-1201-49d6-88f3-56260c8d1bd0">
Rather then using *-notation, we can use syntax similar to the spec
`$runelem(elem*[0], 0) ... $runelem(elem*[$(n-1)], $(n-1))` where `n = |elem*|`

Reply:
Yeah, at this point I wonder if we shouldn't have a version of iteration with an index variable, e.g., `exp^(i<n)`. Then this could be expressed as
```
(runelem_(i, elem))^(i<n)
```
when given `elem^n`. But maybe that's going too far?

Note that the spec already defines an auxiliary `concat` operator for flattening (see [here](https://webassembly.github.io/spec/core/syntax/conventions.html#auxiliary-notation)).
A monomorphic version of that can already be defined in the DSL. For example, this works:
```
def $concat(nat**) : nat*
def $concat(epsilon) = epsilon
def $concat(n* n'**) = n* $concat(n'**)

def $f(nat) : nat*
def $f(n) = n n n
def $g(nat*) : nat*
def $g(n*) = $concat($f(n)*) $concat($f(n)*)
```

Reply:
Oh, Thanks for the ideas! I'll reflect them.

Reply:
We can also use iteration with an index to express the range
For example, `fa* = $range(|s.FUNC|, $(|s.FUNC| + |func*| - 1))` can be replaced by `fa* = $(|s.FUNC| + i)^(i<|func*|)`.
I think it is more compact.